pkg/actors/engine/engine.go

```diff
@@ -0,0 +1,289 @@
+/*
+Copyright 2024 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"k8s.io/utils/clock"
+
+	"github.com/dapr/dapr/pkg/actors/api"
+	actorerrors "github.com/dapr/dapr/pkg/actors/errors"
+	"github.com/dapr/dapr/pkg/actors/internal/placement"
+	"github.com/dapr/dapr/pkg/actors/reminders"
+	"github.com/dapr/dapr/pkg/actors/table"
+	"github.com/dapr/dapr/pkg/actors/targets"
+	"github.com/dapr/dapr/pkg/api/grpc/manager"
+	diag "github.com/dapr/dapr/pkg/diagnostics"
+	diagutils "github.com/dapr/dapr/pkg/diagnostics/utils"
+	internalv1pb "github.com/dapr/dapr/pkg/proto/internals/v1"
+	"github.com/dapr/dapr/pkg/resiliency"
+	"github.com/dapr/kit/concurrency/fifo"
+	"github.com/dapr/kit/events/queue"
+	"github.com/dapr/kit/logger"
+)
+
+var log = logger.NewLogger("dapr.runtime.actor.engine")
+
+type Interface interface {
+	Call(ctx context.Context, req *internalv1pb.InternalInvokeRequest) (*internalv1pb.InternalInvokeResponse, error)
+	CallReminder(ctx context.Context, reminder *api.Reminder) error
+}
+
+type Options struct {
+	Namespace          string
+	Table              table.Interface
+	Placement          placement.Interface
+	Resiliency         resiliency.Provider
+	Reminders          reminders.Interface
+	GRPC               *manager.Manager
+	IdlerQueue         *queue.Processor[string, targets.Idlable]
+	SchedulerReminders bool
+}
+
+type engine struct {
+	namespace          string
+	schedulerReminders bool
+
+	table      table.Interface
+	placement  placement.Interface
+	resiliency resiliency.Provider
+	reminders  reminders.Interface
+	grpc       *manager.Manager
+
+	idlerQueue *queue.Processor[string, targets.Idlable]
+
+	lock    *fifo.Mutex
+	closeCh chan struct{}
+	clock   clock.Clock
+}
+
+func New(opts Options) Interface {
+	return &engine{
+		namespace:          opts.Namespace,
+		schedulerReminders: opts.SchedulerReminders,
+		table:              opts.Table,
+		placement:          opts.Placement,
+		resiliency:         opts.Resiliency,
+		grpc:               opts.GRPC,
+		idlerQueue:         opts.IdlerQueue,
+		reminders:          opts.Reminders,
+		lock:               fifo.New(),
+		closeCh:            make(chan struct{}),
```

Comment:
I think we need a Shutdown func like:

```
func (e *engine) Shutdown() {
    close(e.closeCh)
}
```
Or we might not need the `closeCh` to begin with. I don't see where we are using and closing the channel.

Reply:
Remoed